documentation/proposals/Proposal - Multi-Backend Input.md

```diff
@@ -11,203 +11,683 @@ Proposal API for backend-agnostic, refactored Input via keyboards, mice, and con
+
+# Usage Examples
+```cs
+IWindowHandlesSource someWindow = null!;
+var inputContext = someWindow.CreateInput();
+inputContext.Gamepads.ThumbstickMove += @event =>
+{
+    Console.WriteLine($"Thumbstick {@event.Index} moved from {@event.OldValue} to {@event.NewValue}");
+};
+var isButtonDown = inputContext.Gamepads.Any(gamepadState => gamepadState.Buttons[JoystickButton.A]);
+```
+```cs
+IWindowHandlesSource someWindow = null!;
+var inputContext = new InputContext();
+inputContext.Backends.Add(someWindow.CreateInputBackend());
+// in future:
+// inputContext.Backends.Add(new OpenXRInputBackend(...));
+```
+
+# Reference Implementation
+
+Similar to Windowing 3.0, a reference implementation will be included in the main `Silk.NET.Input` package which uses the same API or family of APIs as Windowing's reference implementation. This will be exposed by the `InputWindowExtensions` class.
+
+```cs
+public static class InputWindowExtensions
+{
+    public static IInputBackend CreateInputBackend(this WindowHandles window);
+    public static IInputBackend CreateInputBackend(this IWindowHandlesSource window);
+    public static InputContext CreateInput(this WindowHandles window);
+    public static InputContext CreateInput(this IWindowHandlesSource window);
+}
+```
+
+The `CreateInputBackend` will create an instance of the reference implementation for the given `WindowHandles`. The `IWindowHandlesSource` overloads just forward to the `WindowHandles` overload. This is because `ISurface` will implement `IWindowHandlesSource`, so the extension methods will be usable on an `ISurface` without having a hard reference between Windowing and Input.
+
+The `CreateInput` methods simply return an `InputContext` preconfigured with the backend created by `CreateInputBackend` for ease of use.
+
+Please see the Windowing 3.0 proposal for `IWindowHandlesSource` and `WindowHandles`.
+
+# Devices, Backends, and Contexts
+
+Input devices all inherit from a root interface.
+
+```cs
+public interface IInputDevice
+{
+    nint BackendId { get; }
+    string Name { get; }
+}
+```
+
+`BackendId` is an integral identifier for this device unique to this backend **only** - i.e. it is not globally unique, and if you have an input context with multiple backends there may be multiple devices with a particular `BackendId`. It may be a handle to memory, or it may be a generic integral index - this is implementation-defined.
+
+`Name` is a rough description of the input device. Its value is not intrinsically meaningful.
+
+All devices originate from a context.
+
+```cs
+public interface IInputBackend
+{
+    string Name { get; }
+    nint Handle { get; }
+    IReadOnlyList<IInputDevice> Devices { get; }
+    void Update();
+}
+```
+
+`Name` is a rough description of the input backend. Its value is not intrinsically meaningful.
+
+`Handle` is a globally-unique handle to memory for this input backend. The memory this handle actually points to is implementation-defined.
```

Comment:
Is it necessary to specify that it is globally unique? Why is that relevant?
Also why do we specify that this points to memory? This seems to only hold true as much as any integer points to *some* memory. (I'd think for OpenXR it'd not be an actual pointer, it'd just be a handle, like everything else? So It'd still theoretically point to memory, but no one would ever treat it as such)

Reply:
Makes sense!

Reply:
```suggestion
`Id` is a globally-unique integral identifier for this backend.
```